src/plugins/api/index_pattern_endpoints.js

```diff
@@ -0,0 +1,221 @@
+        getMappings(pattern, client),
+        stitchPatternAndMappings
+      ).then(removeDeprecatedFieldProps)
+        .then(function (pattern) {
+          reply(pattern);
+        }, function (error) {
+          reply(handleESError(error));
+        });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns',
+    method: 'POST',
+    handler: function (req, reply) {
+      const client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
+      const mappings = _.omit(_.mapValues(_.indexBy(req.payload.fields, 'name'), (value) => {
+        return value.mapping;
+      }), _.isUndefined);
+      indexPattern.fields = JSON.stringify(_.map(indexPattern.fields, (field) => {
+        return _.omit(field, 'mapping');
+      }));
+
+
+      client.create({
+        index: '.kibana',
+        type: 'index-pattern',
+        id: indexPattern.title,
+        body: indexPattern
+      }).then((patternResponse) => {
+        if (!isWildcard || _.isEmpty(mappings)) {
+          return patternResponse;
+        }
+        else {
+          return client.indices.exists({
+            index: indexPattern.title
+          }).then((matchingIndices) => {
+            if (matchingIndices) {
+              throw Boom.conflict('Cannot create an index template if existing indices already match index pattern');
+            }
+            else {
+              return client.indices.putTemplate({
+                order: 0,
+                create: true,
+                name: patternToTemplate(indexPattern.title),
+                body: {
+                  template: indexPattern.title,
+                  mappings: {
+                    _default_: {
+                      properties: mappings
+                    }
+                  }
+                }
+              });
+            }
+          }).catch((templateError) => {
+            return client.delete({
+              index: '.kibana',
+              type: 'index-pattern',
+              id: indexPattern.title
+            }).then(() => {
+              throw templateError;
+            }, () => {
+              throw new Error(`index-pattern ${indexPattern.title} created successfully but index template creation
+                failed. Failed to rollback index-pattern creation, must delete manually.`);
+            });
+          });
+        }
+      }).then(() => {
+        reply('success').statusCode = 201;
+      }).catch(function (error) {
+        reply(handleESError(error));
+      });
+    }
+  });
+
+  server.route({
+    path: '/api/index-patterns/{id}',
+    method: 'PUT',
+    handler: function (req, reply) {
+      let client = server.plugins.elasticsearch.client;
+      const indexPattern = _.cloneDeep(req.payload);
+      const isWildcard = _.contains(indexPattern.title, '*') || (indexPattern.title.match(/\[.*]/) !== null);
```

Comment:
This is a tough call. In a perfect world we wouldn't even need our own `indexPattern` we'd could just use the template as the single source of truth. The tough bit that no matter what we're going to end up with 3 sources of possible mappings for field *foo*:

1. The indexPattern contains the mapping for *foo* as kibana sees it. 
2. The elasticsearch mapping contains fields that already exist in indices that already exist. 
3. The template. The field as it *might* be in indices that don't exist yet. This is crucial for new data coming in, to make sure fields get mapped correctly.

That said, I'm with @spalger here, I don't see any harm in being explicit.